docs/design-documents/20240430-schema-support.md

```diff
@@ -0,0 +1,411 @@
+for the communication between Conduit and the connectors.
+
+#### Option 1: Stream of commands and responses
+
+This pattern is used in WASM processors. A server (in this case: Conduit)
+listens to commands (in this case: via a bidirectional gRPC stream). A client (
+in this case: a connector) sends a command to either register a schema or fetch
+a schema. Conduit receives the command and replies. An example can be seen below:
+
+```protobuf
+rpc CommandStream(stream Command) returns (stream Response);
+```
+
+For different types of commands and response to be supported, `Command`
+and `Response` need to have a `oneof` field where all the possible commands
+and respective responses are listed:
+
+```protobuf
+message Command {
+    oneof cmd {
+        SaveSchemaCommand saveSchemaCmd = 1;
+        // etc.
+   }
+}
+
+message Response {
+  oneof resp {
+    SaveSchemaResponse saveSchemaResp = 1;
+    // etc.
+  }
+}
+```    
+**Advantages**:
+
+1. No additional connection setup. When Conduit starts a connector process, it
+   establishes a connection. The same connection is used for all communication (
+   e.g. configuring a connector, opening, reading/writing records, etc.)
+2. Connector actions (which are planned for a future milestone) might use the
+   same command-and-reply stream.
+
+**Disadvantages**:
+
+1. A single method for all the operations makes both, the server and client
+   implementation, more complex. In Conduit, a single gRPC method needs to check
+   for the command type and then reply with a response. Then the client (i.e.
+   the connector) needs to check the response type. In case multiple commands
+   are sent, we need ordering guarantees.
+
+#### Option 2: Exposing a gRPC service in Conduit
+
+Conduit exposes a service to work with schemas. Connectors access the service
+and call methods on the service. 
+
+For this work, a connector (i.e. clients of the schema service) needs Conduit's
+IP address and the gRPC port. The IP address can be fetched
+using [peer](https://pkg.go.dev/google.golang.org/grpc/peer#Peer). Conduit can
+send its gRPC port to the connector via the `Configure` method.
+
+A skeleton of the gRPC definition of the service would be:
+
+```protobuf
+syntax = "proto3";
+
+service SchemaService {
+  rpc Create(CreateSchemaRequest) returns (CreateSchemaResponse);
+  rpc Fetch(FetchSchemaRequest) returns (FetchSchemaResponse);
+}
+
+message CreateSchemaRequest {
+  string name = 1;
+  // other fields
+}
+
+message CreateSchemaResponse {
+  string id = 1;
+}
+
+message FetchSchemaRequest {
+  string id = 1;
+}
+message FetchSchemaResponse {}
+```
+
+**Advantages**: 
+
+1. Easy to understand: the gRPC methods, together with requests and responses,
+   can easily be understood from a proto file.
+2. An HTTP API for the schema registry can easily be exposed (if needed).
+
+**Disadvantages**:
+
+1. Changes needed to communicate Conduit's gRPC port to the connector.
```

Comment:
Streams are faster (not much of a problem, since schema changes happen less often)